pkg/kv/kvserver/replica_proposal_buf.go

```diff
@@ -416,12 +418,22 @@ func (b *propBuf) FlushLockedWithRaftGroup(
 	// buffer and registering each of the proposals with the proposer, but we
 	// stop trying to propose commands to raftGroup.
 	var firstErr error
-	for i, p := range buf {
+	defer func() {
+		// Clear buffer.
+		for i := range buf {
+			buf[i] = nil
+		}
+	}()
+
+	// `baseIdx` is the index into `ents` of the first element that still has
+	// to get flushed.
+	// INVARIANT: ents[i] corresponds to the proposal at buf[i].
+	baseIdx := 0
```

Comment:
What are we trying to do?

We have the `buf` with the proposals, and `ents` with at most `len(buf)` proposals (some of them can be skipped). For example:

```
buf / props: 1 2 3 4 5 6 7 8 9
       ents: 1 2 4 6 9
```

The `ents` are proposed in chunks, say: `[1 2 4]`, then `[6 9]`.

We want to annotate the `ents` slices with the corresponding `buf / props`, correct? So, we want the iteration to observe:

```
buf / props: 1 2 x 4 | x 6 x x 9
       ents: 1 2   4 |   6     9
```

Instead of writing a fake entry in `ents`, how about "compacting" `props` along the way? (since they are `nil`-ed out anyway)

So:

```golang
buf := b.arr.asSlice()[:used]
ents := make([]raftpb.Entry, 0, used)
defer nilOutBuf()

nextProp := 0 // INVARIANT: len(ents) == nextProp
for _, p := range buf {
	...
	if weWantToAppendToEnts() {
		ents = append(ents, raftpb.Entry{...})
		buf[nextProp] = p
		nextProp++
	}
	if weWantToSendBatch() {
		err := proposeBatch(ctx, raftGroup, b.p.getReplicaID(), ents, buf[:nextProp])
		...
	}
	...
	err := proposeBatch(ctx, raftGroup, b.p.getReplicaID(), ents, buf[:nextProp])
}
```

Reply:
There are some buglets in this snippet, I think we need to also keep track of where we start. But you get the idea hopefully: instead of `ifWeDontWantToAppendToEnts() { insert fake Entry{} to ents}`, flip the condition to `ifWeWantToAppendToEnts() { also track props }`

Reply:
Also, it might be beneficial to move the invariant maintenance into a `func` next to `buf` and `ents` definition. The loop is quite large, so it's hard to identify the algorithm in it.

Reply:
I basically took this suggestion, but I think we can simplify it? We only flush "everything" on ConfChange, so we can "just" keep a pointer to first and last element in `buf` that is current and update. I don't see why we have to mutate the `buf` slice. Chances are I'm asleep at the wheel, but this PR isn't ready for review again anyway, so hold off on checking while I sleep over it and clean up a bit more.